pkg/snapshot/controllerbase.go

```diff
@@ -0,0 +1,245 @@
+func (c *reconcilerBase) reconcileCompletedRequest(ctx context.Context, configMap *corev1.ConfigMap) error {
+	c.logger.Infof("%s request from ConfigMap %s/%s has been completed", c.kind.ToCapital(), configMap.Namespace, configMap.Name)
+	err := c.reconcileDoneRequest(ctx, configMap)
+	if err != nil {
+		return fmt.Errorf("failed to delete %s request Secret %s/%s: %w", c.kind, configMap.Namespace, configMap.Name, err)
+	}
+	return nil
+}
+
+// reconcileFailedRequest cleans up the failed snapshot/restore request resources.
+func (c *reconcilerBase) reconcileFailedRequest(ctx context.Context, configMap *corev1.ConfigMap) error {
+	c.logger.Errorf("%s request from ConfigMap %s/%s has failed", c.kind.ToCapital(), configMap.Namespace, configMap.Name)
+	err := c.reconcileDoneRequest(ctx, configMap)
+	if err != nil {
+		return fmt.Errorf("failed to delete %s request Secret %s/%s: %w", c.kind, configMap.Namespace, configMap.Name, err)
+	}
+	return nil
+}
+
+// reconcileDeletedRequest deletes the snapshot/restore request Secret and removes the finalizer from the
+// snapshot/restore request ConfigMap.
+func (c *reconcilerBase) reconcileDeletedRequest(ctx context.Context, configMap *corev1.ConfigMap) (retErr error) {
+	// snapshot/restore request ConfigMap deleted, so delete Secret as well
+	c.logger.Infof("%s request ConfigMap %s/%s deleted", c.kind.ToCapital(), configMap.Namespace, configMap.Name)
+
+	err := c.reconcileDoneRequest(ctx, configMap)
+	if err != nil {
+		return fmt.Errorf("failed to delete %s request Secret %s/%s: %w", c.kind, configMap.Namespace, configMap.Name, err)
+	}
+	return nil
+}
+
+// reconcileDoneRequest deletes the snapshot/restore request Secret and removes the finalizer from the
+// snapshot/restore request ConfigMap.
+func (c *reconcilerBase) reconcileDoneRequest(ctx context.Context, configMap *corev1.ConfigMap) (retErr error) {
+	defer func() {
+		if retErr != nil {
+			// an error occurred, don't remove the finalizer
+			return
+		}
+		err := c.removeFinalizer(ctx, configMap)
+		if err != nil {
+			retErr = fmt.Errorf(
+				"failed to remove vCluster %s controller finalizer from the %s request ConfigMap %s/%s: %w",
+				c.kind,
+				c.kind,
+				configMap.Namespace,
+				configMap.Name,
+				err)
+		}
+	}()
+
+	err := c.deleteRequestSecret(ctx, configMap)
+	if err != nil {
+		return fmt.Errorf(
+			"failed to delete %s request Secret %s/%s: %w",
+			c.kind,
+			configMap.Namespace,
+			configMap.Name,
+			err)
+	}
+	return nil
+}
+
+func (c *reconcilerBase) deleteRequestSecret(ctx context.Context, configMap *corev1.ConfigMap) error {
+	namespace := configMap.Namespace
+	name := configMap.Name
+	c.logger.Debugf("Deleting %s request Secret %s/%s", c.kind, namespace, name)
+
+	// find snapshot/restore request secret
+	var secret corev1.Secret
+	secretObjectKey := client.ObjectKey{
+		Namespace: namespace,
+		Name:      name,
+	}
+	err := c.client().Get(ctx, secretObjectKey, &secret)
```

Comment:
Do we need to check if the secret exists before before deleting it? I guess the `Delete()` func will return an `not found` error if it doesn't exists